File: 114.cpp
struct Solution {
    using pr = pair<TreeNode*, TreeNode*>;

    pr flatter(TreeNode *node){
        if (node == nullptr){
            return {nullptr, nullptr};
        }
        pr p = this->flatter(node->left), q = this->flatter(node->right);
        TreeNode *lh = p.first, *lt = p.second, *rh = q.first, *rt = q.second, *ptr = node;
        node->left = nullptr;
        if (lh != nullptr){
            ptr->right = lh;
            ptr = lt;
        }
        if (rh != nullptr){
            ptr->right = rh;
            ptr = rt;
        }
        return {node, ptr};
    }
    inline void flatten(TreeNode* root) {
        this->flatter(root);
    }
};

File: 114.py
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        def flatter(node: TreeNode) -> tuple[TreeNode | None, TreeNode | None]:
            if node is None:
                return (None, None)
            lh, lt = flatter(node.left)
            rh, rt = flatter(node.right)
            node.left = None
            ptr = node
            if lh is not None:
                ptr.right = lh
                ptr = lt
            if rh is not None:
                ptr.right = rh
                ptr = rt
            return (node, ptr)
        flatter(root)
